kratos/utilities/function_parser_utility.h

```diff
@@ -0,0 +1,261 @@
+    bool DependsOnSpace()
+    {
+        return mDependsOnSpace;
+    }
+
+    /**
+     * @brief This method returns the function body
+     * @return The function body
+     */
+    std::string FunctionBody();
+
+    /**
+     * @brief This method rotates and calls the evaluation function
+     * @param x The x coordinate
+     * @param y The y coordinate
+     * @param z The z coordinate
+     * @param t The time variable
+     * @param X The initial x coordinate
+     * @param Y The initial y coordinate
+     * @param Z The initial z coordinate
+     */
+    double RotateAndCallFunction(
+        const double x,
+        const double y,
+        const double z,
+        const double t,
+        const double X = 0.0,
+        const double Y = 0.0,
+        const double Z = 0.0
+        );
+
+    /**
+     * @brief This calls the evaluation function
+     * @param x The x coordinate
+     * @param y The y coordinate
+     * @param z The z coordinate
+     * @param t The time variable
+     * @param X The initial x coordinate
+     * @param Y The initial y coordinate
+     * @param Z The initial z coordinate
+     */
+    double CallFunction(
+        const double x,
+        const double y,
+        const double z,
+        const double t,
+        const double X = 0.0,
+        const double Y = 0.0,
+        const double Z = 0.0
+        );
+
+    ///@}
+
+private:
+
+    ///@name Member Variables
+    ///@{
+
+    std::unordered_map<std::string, double>  mNameSpace;   /// The variables considered on the function
+    te_expr* mpTinyExpr = nullptr;                         /// The function parser
+    std::string mFunctionBody;                             /// The function body
+
+    bool mDependsOnSpace = true;                 /// If it depends on space
+    bool mUseLocalSystem = false;                /// If we use a local system
+    BoundedMatrix<double, 3, 3> mRotationMatrix; /// The rotation matrix
+    array_1d<double, 3> mCenterCoordinates;      /// The center of coordinates
+
+    ///@}
+    ///@name Private Operations
+    ///@{
+
+    /**
+     * @brief This method initializes the parser classes
+     */
+    void InitializeParser();
+
+    ///@}
+}; /// GenericFunctionUtility
+
+/**
+ * @class FunctionParser
+ * @ingroup KratosCore
+ * @brief This class parses a string as a std::function
+ * @author Vicente Mataix Ferrandiz
+ */
+class FunctionParser
+{
+public:
+    /**
+     * @brief This method parses a string into a std::function
+     * @details This is the version with 3 arguments (x,y,z)
+     * @param rFunctionBody The string defining the function
+     */
+    static std::function<double(const double, const double, const double)> GenerateFunctionSpace(const std::string& rFunctionBody)
+    {
+        std::function<double(const double, const double, const double)> function = [rFunctionBody](const double x, const double y, const double z) -> double
+        {
+            GenericFunctionUtility parser(rFunctionBody);
```

Comment:
here you parse the function every time you generate it. That should be very slow...

Reply:
Yes, but you are supposed to create it just once, not everytime 

BTW, this is brand new, not used anywhere

Reply:
as i understand the current code, it will parse the rFunctionBody every time you call "function".

also note that right now rFunctionBody  is passed BY COPY to the lambda.

Reply:
The only way to solve this is make it is removing the static methods

Reply:
well, this is a blocker...

Reply:
actually you are not right, you can pass the parsed function as a parameter to the std::function

Reply:
Indeed, but I need to change the code a little bit